---
title: "Exercise_04
output: html_notebook
---
1. Gene CD33. Use grep to find the index of the important gene CD33
among the list of characters golub.gnames. For each test below for-
mulate the null hypothesis, the p-value and your conclusion.
(a) Test the normality of the ALL and AML expression values.
Q-Q Plot indicates a close to normal distribution distribution of gene expression data, 
Shapiro-Wilkinson and Anderson Darling Test show a p-value well above 0.05-> the H0 is not rejected
and the data is very likely normal distributed. Note that for ALL the data is much closer to 
normal distribution than for AML, which has extreme outliers

(b) Test for the equality of variances.
F-Test gives ratio of variances 
         0.4604523 
  p-value = 0.1095 H0 is not rejected, AML and ALL Have an equal variance

(c) Test for the equality of the means by an appropriate t-test.
P-value is very small, H0:Equal means can be rejected. 
(And mean is very different too!)

(d) Is the experimental effect strong?
t = -6.7878 indicates a very strong effect 

```{r}
data(golub,package="multtest")
#----create factor----
gol.fac <- factor(golub.cl,levels=0:1, labels= c("ALL","AML"))
#----find index----
grep("CD33",golub.gnames[,2])
#a)====
#Normality-Test: Shapiro-Wilk and Anderson Darling test
shapiro.test(golub[808, gol.fac=="ALL"])
shapiro.test(golub[808, gol.fac=="AML"])

library(nortest)
ad.test(golub[808,gol.fac=="ALL"])
ad.test(golub[808,gol.fac=="AML"])
# perform Q-Q plot to see distribution against the normal distribution
qqnorm(golub[808,gol.fac=="ALL"])
qqline(golub[808,gol.fac=="ALL"])
qqnorm(golub[808,gol.fac=="AML"])
qqline(golub[808,gol.fac=="AML"])

#b)====
#Equality of Variances Test: F-Test
var.test(golub[808,] ~gol.fac)

#c)====
#T-test for equality of means: Z-Test not applicable as sd is not known
#Use the two sample t-Test as two groups of patience are available
t.test(golub[808,] ~gol.fac)


```

2. Gene ”MYBL2 V-myb avian myeloblastosis viral oncogene homolog-
like 2” has its expression values in row 1788.
(a) Use a boxplot to construct a hypothesis about the experimental
effect.

The Boxplot shows no outliers, the distribution of the quartiles is about equal,
both indicate for a normally distributed set which has little to no effect on experimental outcome

(b) Test for the equality of means by an appropriate t-test.
```{r}

#a====
data(golub,package="multtest")
#----create factor----
gol.fac <- factor(golub.cl, levels=0:1, labels=c("ALL","AML"))
#----find index----
grep("MYBL2",golub.gnames[,2])
#----extract expression values----
golub[1788,]
#----contruct quantiles----
qqnorm(golub[1788,])
qqline(golub[1788,])
#----boxplot----
boxplot(golub[1788,]~gol.fac)

#----t-test for equality of means----
t.test(golub[1788, ]~gol.fac,var.equal=TRUE)

```

3. HOXA9. Gene ”HOXA9 Homeo box A9” with expression values in row
1391, can cause leukemia (Golub et al., 1999).
(a) Test the normality of the expression values of the ALL patients.
Reject, because of the p-value of the SHapiro Wilk Test is way below 0,05
(b) Test for the equality of means by an appropriate t-test.
Rejected, as the Wilcox-Test for non-normally distributed data with equal variance 
gives a too small p-value for the hypothesis to be accepted.
```{r}
data(golub,package="multtest")
#----create factor----
gol.fac <- factor(golub.cl, levels=0:1, labels=c("ALL","AML"))
#----find index----
grep("HOXA9",golub.gnames[,2])
#----extract expression values----
golub[1391,]

#====test for normality====
#Shapiro-Wilk 
shapiro.test(golub[1391, gol.fac=="ALL"])
#====test for equality of means====
wilcox.test(golub[1391,] ~gol.fac)


```

4. Zyxin. On NCBI there are various cDNA clones of zyxin.
(a) Find the accession number of cDNA clone with IMAGE:3504464.
BC002323.2.
(b) Test whether the frequencies of the nucleotides are equal for each
nucleic acid.
They are not. Chi-Squared Test has a very very small p-value, way below 0.05
(c) Test whether the frequencies of ”X94991.1” can be predicted by
the probabilities of the cDNA sequence ”BC002323.2”.
Chi-Squared Test: Yes it can, confidence at 99.88 %

```{r}
#----read from genbank----
library(ape)
library(annotate)
library(dplyr)

freq_Zyxin<-table(read.GenBank(c("3504464"),as.character=TRUE))

#====Test for multiple probabilities====
#Use Chi-Squared!
chisq.test(freq_Zyxin)
#====c)====

x <- base.freq(read.GenBank(c("BC002323.2")),freq = TRUE, all= FALSE )
x
y <- base.freq(read.GenBank(c("X94991.1")), freq = TRUE, all = FALSE)
y

chisq.test(x, p=y/sum(y))

```
5. Gene selection. Select the genes from the golub data with smallest
two-sample t-test values for which the ALL mean is greater than the
AML mean. Report the names of the best ten. Scan the Golub (1999)
article for genes among the ten you found and discuss their biological
function briefly.

I could find none. There was no such thing in the pdf :*(

Hypothesis: ALL mean is greater than AML mean
1. Import Golub
2. perform two-sample t-test 
3. Check where ALL > AML
4. Sort and extract top 10 
5. Scan Golub article for these 10
6. get biological function
```{r}
library("here")

data(golub, package="multtest")
gol.fac <- factor(golub.cl, levels = 0:1, labels = c("ALL","AML"))

all_genes <- apply(golub, 1, 
                   function(x) t.test (x~gol.fac, 
                                       alternative = c("greater")
                                       )$p.value)
topten_genes <- golub.gnames[order(all_genes)[1:10],2]
topten_genes

library(pdfsearch)
topten_genes[1:10]
#file <- system.file('pdf', here("Ressources","Molecular_classification_of_cancer_class_discovery.pdf"), package = 'pdfsearch')

result <- keyword_search(here("Ressources","Molecular_classification_of_cancer_class_discovery.pdf"), 
            keyword = c("X-LINKED HELICASE II"), path = TRUE, surround_lines = 1)

head(result)
head(result$line_text, n = 1)


```
6. Antigenes. Antigenes play an important role in the development of
cancer. Order the antigenes according to their p-values from the Welch
two-sample t-test with respect to gene expression values from the ALL
and AML patients of the Golub et al. (1999) data.
```{r}
library(multtest); data(golub)
gol.fac <- factor(golub.cl,levels=0:1, labels= c("ALL","AML"))
#T Test to all patients
pt <- apply(golub, 1, function(x) t.test(x ~ gol.fac)$p.value)
#Seeking only those with antigen / index
index <-agrep("^antigen",golub.gnames[,2])
#New Object containing only antigens
golub.index<-golub[index,]
#tTest Probability of antigens in one object
pt.index<-pt[index]
#Get Names to index 
golub.gnames.index<-golub.gnames[index,]

golub.gnames.index[order(pt.index)[1:length(index)],2]
```
7. Genetic Model. A certain genetic model predicts that four phenotypes
occur in ration 9:3:3:1. In a certain experiment the offspring is observed
with frequencies 930, 330, 290, 90. Do the data confirm the model?
Chi Squared Test wird benoetigt. 
9/16:3/16:3/16:1/16
vs 
930/SUM:330/SUM:....
Accepted, because p-value is above 0.05

```{r}
x <- c(930,330,290,90)
pi <- c(9/16, 3/16, 3/16,1/16)
chisq.test(x, p=pi)


```

8. Comparing two genes. Consider the gene expression values in row 790
and 66 of the Golub et al. (1999) data.
(a) Produce a boxplot for the ALL expression values and comment on
the differences. Are there outliers?

Yes, for 790 there are some extreme outliers to the top (overexpressed ?)

(b) Compute the mean and the median for the ALL gene expression
values for both genes. Do you observed difference between genes?

Yes.
Mean of 66 is positive, mean of 790 negative. |mean| is almost the same. 

(c) Compute three measures of spread for the ALL expression values
for both genes. Do you observe difference between genes?

Measures of spread: 
Standard deviation (SD)
Interquartile Range (IQR)
median absolute deviation (MAD)

yes, there are very clear differences. 
The MAD (mean absolute deviation) shows the spread of data by being the average of distances to the mean. 
The bigger the MAD the more spread out data is, 66 is far more spread out than 790, following this value.

The IQR is the range between the first and the third quartile. Divide it by 1.349 to get an estimate of the standard deviation. 
Here, the range (biggest-smallest value) of the IQR is much larger for 66, indicating that the mid data is spread out further than with 790. 

Variance:
The Variance is the average of the squared differences from the mean. Small variance indicates a smaller spread. 

Standard deviation: The average distance of the datapoints to the mean. The SD is the square root of the variance. (Outliers have a much bigger effect than for MAD)
Small SD indicates smaller spread. IQR/1.349 approximates the sd. 


[1] "All66: sd, mean, IQR, mad"
[1] 0.4408175
[1] 1.182503
[1] 0.57737
[1] 0.4523413
[1] "ALL_790: sd, mean, IQR, mad"
[1] 0.4809953
[1] 1.182503
[1] 0.20667
[1] 0.1864814

These data indicate that ALL_66 had a much more outspread mid 50% of data, with the MAD vAlue being higher than in the ALL_790 set too. 
However the sd are about the same, as well as their mean. This is only plausible if the 790 dataset has extreme outliers which allow for equal mean, but its data being rather centered around the mean in general. Looking at the boxplots, this indeed is confirmed, ALL_66 is more widespread in its mid 50% and has no extreme outliers, while ALL790 has.

(d) Test by Shapiro-Wilk and Anderson-Darling the normality for the
ALL gene expression values for both genes.

The shapiro Wilk test (based on degree of linearity in Q-Q Plot) indeed indicates a normal distribution for all 66, while all 790 has to be rejected.
The Anderson Darling Test, based on the distribution of data, too accepts normality for all66 and rejects it for all790, which means that the outliers of 790 must have an extreme influence on their distribution.
```{r}
data(golub, package="multtest")

gol.fac <-factor(golub.cl,levels=0:1,labels=c("ALL","AML"))
all_66 <- golub[66, gol.fac=="ALL"]
all_790 <-golub[790,gol.fac=="ALL"]

boxplot(all_66, all_790)
#b====
mean(all_66)
mean(all_790)
median(all_66)
median(all_790)
#c)====
print("All66: sd, mean, IQR, mad")
sd(all_66)
mean(all_66)
IQR(all_66)
mad(all_66)

IQR(all_66)/1.349

print("ALL_790: sd, mean, IQR, mad")

sd(all_790)
mean(all_66)
IQR(all_790)
mad(all_790)
IQR(all_790)/1.349

#d)====
shapiro.test(all_66)
shapiro.test(all_790)

library(nortest)
ad.test(all_66)
ad.test(all_790)

```

9. Normality tests for gene expression values of the Golub et al. (1999)
data. Perform the Shapiro-Wilk normality test separately for the ALL
and AML gene expression values. What percentage passed the normal-
ity test separately for the ALL and the AML gene expression values?
What percentage passes both testes?
ALL PASS: 58,28%
AML PASS: 78,56%

BOTH: 49,72%

```{r}
data(golub, package = "multtest")
gol.fac <- factor(golub.cl,0:1, labels=c("ALL","AML"))

allsh <- apply(golub[,gol.fac=="ALL"], 1, function(x) shapiro.test(x)$p.value)
amlsh <- apply(golub[,gol.fac=="AML"], 1, function(x) shapiro.test(x)$p.value)

100 * sum(allsh>0.05)/length(allsh)
#SUM creates a vector with TRUE or FALSE. Since TRUE and FALSE are 1 and 0, the sum over those values equals the amount of values greater than specified)
100 * sum(amlsh>0.05)/length(amlsh)
#Percentage passing both tests:
100 * sum(allsh>0.05 & amlsh > 0.05)/length(amlsh)

```
10. Two-sample tests on gene expression values of the Golub et al. (1999)
data.
(a) Perform the two-sample Welch t-test and report the names of the
ten genes with the smallest p-values.
(b) Perform the Wilcoxon rank test and report the names of the ten
genes with the smallest p-values.
```{r}
library(dplyr)
data(golub, package = "multtest")
gol.fac <- factor(golub.cl,0:1, labels=c("ALL","AML"))

#a)====
welch_test <- apply(golub,1, function(x) t.test(x~gol.fac)$p.value )
o <- order(welch_test,decreasing=FALSE)
golub.gnames[o[1:10],2]

#dplyr nor useable, as object has to exist beforehand

#b)====

wilcox_test <- apply(golub,1, function(x) wilcox.test(x~gol.fac)$p.value)
o <- order(wilcox_test,decreasing=FALSE)
golub.gnames[o[1:10],2]

```
11. Biological hypotheses. Suppose that the probability to reject a biolog-
ical hypothesis by the results of a certain experiment is 0.05. Suppose
that the experiment is repeated 1000 times.

(a) How many rejections do you expect.
n=1000, p=0.05, n*p=50

(b) What is the probability of less than 10 rejections?
Binomial Distribution; either 0 or 1 are the results -> pbinom(9,1000,0.05)
5.242679e-13

(c) What is the probability of more than 5 rejections?
1

(d) What is the probability that the number of rejections is between
two and eight?
8.837404e-14

```{r}
#a)====
#n=1000, a=0.05 np=50
#b)====
pbinom(9,1000,0.05)
#c)====
1-pbinom(5,1000,0.05)
#solution from paper:
sum(dbinom(6:1000,1000,.05))
#d)====
sum(dbinom(2:8,1000,0.05))


```

12. Programming some tests.
(a) Program the two-sample t-test with equal variances and illustrate
it with the expression values of row 1024 the of Golub et al. (1999)
data.
(b) The value of W in the two-sample Wilxoxon test equals the sum
of the ranks of Group 1 minus n(n + 1)/2, where n is the number
of gene expression values in Group 1. Program this and illustrate
it with the expression values of row 1024 of Golub et al. (1999)
data.
(c) The value of W in the two-sample Wilxoxon test equals the num-
ber of values x i > y j , where x i , y j are values from Group 1 and
2, respectively. Program this and illustrate it with the expression
values of row 1024 of Golub et al. (1999) data.
```{r}
data(golub, package = "multtest")
gol.fac <- factor(golub.cl, 0:1, labels = c("ALL","AML"))

t.test(golub[1024,] ~gol.fac)
x <- golub[1042,gol.fac=="ALL"]
n <- length(x)
y <- golub[1042,gol.fac=="AML"]
m <- length(y)
t <- (mean(x)-mean(y))/sqrt(var(x)/n + var(y)/m)
v <- (var(x)/n + var(y)/m)^2/( (var(x)/n)^2/(n-1) + (var(y)/m)^2/(m-1)+2*pt(-abs(t),v))
                               
mean(x) - mean(y) + qt(0.025,v)* sqrt(var(x)/n + var(y)/m)

mean(x) - mean(y) + qt(0.975,v)* sqrt(var(x)/n + var(y)/m)

```
